pulp_rpm/app/tasks/copy.py

```diff
@@ -102,48 +76,56 @@ def copy_content(source_repo_version_pk, dest_repo_pk, criteria, content_pks, de
+        if entry.get("content"):
+            content_filter = Q(pk__in=entry.get("content"))
+        else:
+            content_filter = Q()
 
-    if not dependency_solving:
-        content_to_copy = _filter_content(source_repo_version.content, criteria, content_pks)
-        content_to_copy |= find_children_of_content(content_to_copy, source_repo_version)
+        if not dependency_solving:
+            content_to_copy = source_repo_version.content.filter(content_filter)
+            content_to_copy |= find_children_of_content(content_to_copy, source_repo_version)
 
-        with dest_repo.new_version() as new_version:
-            new_version.add_content(content_to_copy)
+            with dest_repo.new_version() as new_version:
+                new_version.add_content(content_to_copy)
 
-        return
+            continue
 
-    # Dependency Solving Branch
-    # =========================
-    content_to_copy = {}
+        # Dependency Solving Branch
+        # =========================
+        content_to_copy = {}
 
-    # TODO: add lookaside repos here
-    repo_mapping = {source_repo_version: dest_repo_version}
-    libsolv_repo_names = {}
+        # TODO: add lookaside repos here
+        repo_mapping = {source_repo_version: dest_repo_version}
+        libsolv_repo_names = {}
 
-    solver = Solver()
+        solver = Solver()
 
-    for src in repo_mapping.keys():
-        repo_name = solver.load_source_repo(src)
-        libsolv_repo_names[repo_name] = src
+        for src in repo_mapping.keys():
+            repo_name = solver.load_source_repo(src)
+            libsolv_repo_names[repo_name] = src
 
-        content = _filter_content(src.content, criteria, content_pks)
-        children = find_children_of_content(content, src)
-        content_to_copy[repo_name] = content | children
+            content = src.content.filter(content_filter)
+            children = find_children_of_content(content, src)
+            content_to_copy[repo_name] = content | children
 
-    for tgt in repo_mapping.values():
-        solver.load_target_repo(tgt)
+        for tgt in repo_mapping.values():
+            solver.load_target_repo(tgt)
 
-    solver.finalize()
+        solver.finalize()
 
-    content_to_copy = solver.resolve_dependencies(content_to_copy)
+        content_to_copy = solver.resolve_dependencies(content_to_copy)
 
-    for from_repo, units in content_to_copy.items():
-        src_repo_version = libsolv_repo_names[from_repo]
-        dest_repo_version = repo_mapping[src_repo_version]
-        with dest_repo_version.repository.new_version() as new_version:
-            new_version.add_content(Content.objects.filter(pk__in=units))
+        for from_repo, units in content_to_copy.items():
+            src_repo_version = libsolv_repo_names[from_repo]
+            dest_repo_version = repo_mapping[src_repo_version]
+            with dest_repo_version.repository.new_version() as new_version:
```

Comment:
We probably need to structure this so that all repository versions are "finalized" before they are actually committed to the database, that way we're not completing one repo version and then experiencing a failure with the next repo version in the copy.